### PR TITLE
Update path for Port Channel configuration. #48

### PR DESCRIPTION
**Pull request by rahul-1475**
Issues:

- If we try to change the interface mode of a portchannel from "trunk" to "access", it is required to first delete trunk config from port channel, but from "access" to "trunk" restconf path doesn't require to delete "access" configuration. Test if it is the similar case with gNMI as well ? If so we can mitigate this in ORCA i.e. while making a port channel to be a "access" interface we first delete "trunk" config (only if there was any trunk config already.)
- Also while updating access vlan number on a portchannel which is already a in access mode, resconf requires to first delete the access config then apply new config with new vlan. Test if it is the similar case with gNMI? If so we also mitigate the 2 steps into one in ORCA i.e. If user requests to update access vlan number on a portchannel which is in interface mode "access" first delete the access config (only when there is already a access vlan assigned anf interface mode is "access") then apply new config.
-The API should take interface mode (enum) as an input and list of vlans, in case of access mode vlan list will have only one element .

Fixes:

- changes to for input to take if_mode and vland_ids.
- code changes to updates, access to trunk, tunk to access,  updating access valn with new vlan, updating tunk vlans.
- added new test cases to test switch between different access modes.
- modified exting test cases to use if_mode and vlan_ids as input.

Note: while switching deom tunk to access or updting access vlan. Existing vlans are removed beacuse inorder to update access vlan exting vlan must be deleted and upting from tunk to access trunk vlan needs to be deleted.

### CODE REVIEW

**Comment by rahul-1475** [comment]
orca_backend pr: https://github.com/STORDIS/orca_backend/pull/54

File: orca_nw_lib/port_chnl_gnmi.py [hunk 1]
```diff
@@ -457,33 +449,79 @@ def get_port_channel_vlan_memebers_path(port_channel_name: str):
...
     Returns:
         The GNMI update request for configuring VLAN members on the specified port channel.
     """
-    vlan_config_members = {}
-    if trunk_vlans:
-        # checking input format change into list of ids
-        vlan_config_members["trunk-vlans"] = format_and_get_trunk_vlans(trunk_vlans)
-    if access_vlan:  # checking if access vlan is empty
-        vlan_config_members["access-vlan"] = access_vlan
+    if if_mode == IFMode.ACCESS:
```
Comment (kamalkrbh) [inline orca_nw_lib/port_chnl_gnmi.py:466]: Before sending a gnmi request better check nullability of of ifmode and vlanids(here also check empty array) Also log respecive error message.

File: orca_nw_lib/port_chnl_gnmi.py [hunk 2]
```diff
@@ -457,33 +449,79 @@ def get_port_channel_vlan_memebers_path(port_channel_name: str):
...
+        # Remove all existing VLANs before adding new ones.
+        # This is necessary because updating an access VLAN requires the existing access VLAN to be deleted first.
+        # Additionally, switching from trunk to access mode requires the deletion of trunk VLANs.
+
+        delete_all_port_channel_member_vlan_from_device(device_ip, port_channel_name)
+        return create_req_to_add_access_vlan(
+            port_channel_name=port_channel_name, vlan_id=vlan_ids[0]
```
Comment (kamalkrbh) [inline orca_nw_lib/port_chnl_gnmi.py:473]: Document why retrieving only the first element of the array.

File: orca_nw_lib/port_chnl_gnmi.py [hunk 3]
```diff
@@ -457,33 +449,79 @@ def get_port_channel_vlan_memebers_path(port_channel_name: str):
...
-        # checking input format change into list of ids
-        vlan_config_members["trunk-vlans"] = format_and_get_trunk_vlans(trunk_vlans)
-    if access_vlan:  # checking if access vlan is empty
-        vlan_config_members["access-vlan"] = access_vlan
+    if if_mode == IFMode.ACCESS:
+        # Remove all existing VLANs before adding new ones.
+        # This is necessary because updating an access VLAN requires the existing access VLAN to be deleted first.
+        # Additionally, switching from trunk to access mode requires the deletion of trunk VLANs.
+
+        delete_all_port_channel_member_vlan_from_device(device_ip, port_channel_name)
```
Comment (kamalkrbh) [inline orca_nw_lib/port_chnl_gnmi.py:471]: should not delete the vlan members implicitely, Instead, user will automatically receive a message while deleting the channel that there are already member vlan's present.

File: orca_nw_lib/port_chnl.py [hunk 4]
```diff
@@ -292,26 +305,26 @@ def get_port_chnl_members(device_ip: str, port_chnl_name: str, ifname: str = Non
         ]
 
 
-def add_port_chnl_vlan_members(device_ip: str, chnl_name: str, trunk_vlans: list[int] = None, access_vlan: int = None):
+def add_port_chnl_vlan_members(device_ip: str, chnl_name: str, if_mode: IFMode, vlan_ids: list[int] = None):
```
Comment (kamalkrbh) [inline orca_nw_lib/port_chnl.py:308]: vlan_ids can not be None

**Comment by kamalkrbh** [comment]
@rahul-1475 - 

- [ ] Please Have a look at the commets provided.
- [ ] Merge with main and resolve conflicts.
- [ ] All the tests should pass again.

File: orca_nw_lib/port_chnl_gnmi.py [hunk 5]
```diff
@@ -457,33 +449,79 @@ def get_port_channel_vlan_memebers_path(port_channel_name: str):
...
+        return create_req_to_add_access_vlan(
+            port_channel_name=port_channel_name, vlan_id=vlan_ids[0]
+        )
+    if if_mode == IFMode.TRUNK:
```
[the remark this answers is not in the document]
Comment (rahul-1475) [reply orca_nw_lib/port_chnl_gnmi.py]: added elif conditiona and else block to log error message and raise error

File: orca_nw_lib/port_chnl_gnmi.py (hunk 2 above)
Comment (rahul-1475) [reply orca_nw_lib/port_chnl_gnmi.py:477]: added comment why retriving firest element of array

File: orca_nw_lib/port_chnl.py (hunk 4 above)
Comment (rahul-1475) [reply orca_nw_lib/port_chnl.py]: removed none as default value

**Pull request merged**